### PR TITLE
feat: Alby integration

### PR DESCRIPTION
**Pull request by rolznz**
## What does this PR do?

Alby Implementation for https://github.com/calcom/cal.com/pull/11347

This enables Cal.com users to make setup paid meetings and accept sats from their guests. Guests must pay in advance to book the meeting.

Features:
- Connect with your existing Alby account with or create a new one with a few clicks
- Pay for meetings by scanning, clicking, copying, or with a WebLN extension
- Sats - USD conversion on hover

This PR also introduces some additional functionality which needs to be reviewed:
- support for custom currencies / currency symbols
- App cards now display if an app is setup correctly and has a link to the setup page
- Add a way to navigate between setup page and app store entry
- Payment webhook handling extracted into a new lib function
- Replaced getStaticProps with getServerSideProps on app setup pages in order to access app credentials (maybe there is a better way and this change can be undone)

https://www.loom.com/share/75dfc1d394bf4756a4520523afb913fd?sid=37430206-7e7d-4f85-af37-a454f144c9cc

## Requirement/Documentation

- If there is a requirement document, please, share it here.
- If there is ab UI/UX design document, please, share it here.

N/A

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Chore (refactoring code, technical debt, workflow improvements)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How should this be tested?

Manually

- Are there environment variables that should be set? yes: Alby OAuth client ID and secret (see .env.example)
- What are the minimal test data to have?
- What is expected (happy path) to have (input and output)? Booking successfully completed after payment
- Any other important info that could help to test that PR: use a tunnel (e.g. ngrok) so that payment webhooks work

## Mandatory Tasks

- [x] Make sure you have self-reviewed the code. A decent size PR without self-review might be rejected.

## Checklist

- My code doesn't follow the style guidelines of this project
- I haven't added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by PeerRich** [comment]
amazing work! exciting to merge this week and install myself. huge fan of alby

File: .env.example [hunk 1]
```diff
@@ -230,6 +230,12 @@ AUTH_BEARER_TOKEN_VERCEL=
 E2E_TEST_APPLE_CALENDAR_EMAIL=""
 E2E_TEST_APPLE_CALENDAR_PASSWORD=""
 
+# ALBY
+# Used for the Alby payment app / receiving Alby payments
+# Get it from: https://getalby.com/developer/oauth_clients
+NEXT_PUBLIC_ALBY_CLIENT_ID=
+NEXT_PUBLIC_ALBY_CLIENT_SECRET=
+
```
Comment (alannnc) [inline .env.example:238]: moving this to .env.appStore.example

Locally I don't know why but in my setup I still have to set this on .env

File: apps/web/components/apps/AppPage.tsx [hunk 2]
```diff
@@ -89,6 +89,8 @@ export const AppPage = ({
 
   const [existingCredentials, setExistingCredentials] = useState<number[]>([]);
   const [showDisconnectIntegration, setShowDisconnectIntegration] = useState(false);
+  // FIXME: remove hardcoding
+  const showSetupIntegration = slug === "alby";
```
Comment (alannnc) [inline apps/web/components/apps/AppPage.tsx:93]: We can omit this for now, since you have the "This app has not been setup yet" and action button there I think it works great now.

File: packages/app-store/alby/components/AlbyPriceComponent.tsx [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+      const unformattedFiatValue = await fiat.getFiatValue({ satoshi: price, currency: "USD" });
+      setFiatValue(`$${unformattedFiatValue.toFixed(2)}`);
+    })();
+  }, [price]);
+
+  return (
+    <Tooltip content={fiatValue}>
+      <div className="inline-flex items-center justify-center">
+        {displaySymbol && <SatSymbol className="h-4 w-4" />}
+        {price}
```
Comment (alannnc) [inline packages/app-store/alby/components/AlbyPriceComponent.tsx:25]: Should we add the word "sats" here?

Comment (alannnc) [reply packages/app-store/alby/components/AlbyPriceComponent.tsx:25]: image

File: packages/app-store/_components/AppCard.tsx [hunk 4]
```diff
@@ -111,8 +112,23 @@ export default function AppCard({
       </div>
       <div ref={animationRef}>
         {app?.isInstalled && switchChecked && <hr className="border-subtle" />}
+
         {app?.isInstalled && switchChecked ? (
-          <div className="p-4 pt-5 text-sm [&_input]:mb-0 [&_input]:leading-4">{children}</div>
+          app.isSetup ? (
```
Comment (alannnc) [inline packages/app-store/_components/AppCard.tsx:117]: Nice job here!

File: packages/app-store/alby/api/add.ts [hunk 5]
```diff
@@ -1,4 +1,47 @@
-import { createDefaultInstallation } from "@calcom/app-store/_utils/installation";
+import type { NextApiRequest, NextApiResponse } from "next";
+
+import prisma from "@calcom/prisma";
+
+import config from "../config.json";
+
+// FIXME: is a custom handler really needed?
```
Comment (alannnc) [inline packages/app-store/alby/api/add.ts:7]: @hariombalhara I would leave this to you since you have more knowledge of appStore

File: packages/app-store/alby/components/AlbyPaymentComponent.tsx [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+  }
+  const paymentRequest = parsedData.data.invoice.paymentRequest;
+
+  return (
+    <div className="mb-4 mt-8 flex h-full w-full flex-col items-center justify-center gap-4">
+      <PaymentChecker {...props.paymentPageProps} />
+      {isPaying && <Spinner className="mt-12 h-8 w-8" />}
+      {!isPaying && (
+        <>
+          {/* <p className="break-all">Payment {JSON.stringify(payment)}</p> */}
```
Comment (alannnc) [inline packages/app-store/alby/components/AlbyPaymentComponent.tsx:58]:
```suggestion
```

File: packages/app-store/alby/components/AlbyPaymentComponent.tsx [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+    </div>
+  );
+};
+
+type PaymentCheckerProps = PaymentPageProps;
+
+function PaymentChecker(props: PaymentCheckerProps) {
+  // This effect checks if the booking status has changed to "ACCEPTED"
+  // then reload the page to show the new payment status
+  // FIXME: subscribe to the exact payment instead of polling bookings
```
Comment (alannnc) [inline packages/app-store/alby/components/AlbyPaymentComponent.tsx:128]: We can change this for a TODO: probably enabling payments for trpc and then you actually listen for any success there.

File: packages/trpc/server/routers/loggedInViewer/integrations.handler.ts [hunk 8]
```diff
@@ -169,6 +169,8 @@ export const integrationsHandler = async ({ ctx, input }: IntegrationsOptions) =
         invalidCredentialIds,
         teams,
         isInstalled: !!userCredentialIds.length || !!teams.length || app.isGlobal,
+        // FIXME: remove hardcoding and add per-app validation
+        isSetup: !!(credential?.key as { account_id: string })?.account_id || app.slug !== "alby",
```
Comment (alannnc) [inline packages/trpc/server/routers/loggedInViewer/integrations.handler.ts:173]: We can leave this for later I think. The idea we can probably validate using zod schemas.

**Review by alannnc — COMMENTED** [review]
I tested and it worked great, made some little changes. But since it's touching other payment providers I still need to test those too. Also if possible remove every fix me or change them for todo later.

File: packages/app-store/alby/components/AlbyPriceComponent.tsx (hunk 3 above)
Comment (rolznz) [reply packages/app-store/alby/components/AlbyPriceComponent.tsx:25]: @alannnc sounds good to me.

File: .env.appStore.example [hunk 9]
```diff
@@ -125,4 +125,11 @@ SALESFORCE_CONSUMER_SECRET=""
 ZOHOCRM_CLIENT_ID=""
 ZOHOCRM_CLIENT_SECRET=""
 
+# ALBY
+# Used for the Alby payment app / receiving Alby payments
+# Get it from: https://getalby.com/developer/oauth_clients
+# Set callbackUrl to /api/integrations/alby/alby-webhooks
```
Comment (rolznz) [inline .env.appStore.example:131]: it should be `webhook` not `webhooks` - I will update in a follow up commit. 

Thanks for updating the comment

File: packages/app-store/_components/AppCard.tsx (hunk 4 above)
Comment (rolznz) [reply packages/app-store/_components/AppCard.tsx:117]: Thanks!

File: packages/app-store/alby/api/add.ts (hunk 5 above)
Comment (rolznz) [reply packages/app-store/alby/api/add.ts]: I updated the comment. The only thing we want to do here is redirect to the setup page after installing the app.

File: packages/app-store/alby/components/AlbyPaymentComponent.tsx (hunk 7 above)
Comment (rolznz) [reply packages/app-store/alby/components/AlbyPaymentComponent.tsx]: updated, thanks

File: packages/trpc/server/routers/loggedInViewer/integrations.handler.ts (hunk 8 above)
Comment (rolznz) [reply packages/trpc/server/routers/loggedInViewer/integrations.handler.ts:173]: updated comment, thanks

File: packages/app-store/alby/components/AlbyPriceComponent.tsx (hunk 3 above)
Comment (rolznz) [reply packages/app-store/alby/components/AlbyPriceComponent.tsx:25]: Yes, I think that's better and more understandable for users. I've added it now. Thanks!

File: packages/app-store/alby/lib/parseInvoice.ts [hunk 10]
```diff
@@ -9,7 +9,13 @@ export default function parseInvoice(
...
   webhookEndpointSecret: string
-): Invoice {
-  const wh = new Webhook(webhookEndpointSecret);
-  return wh.verify(body, headers) as Invoice;
+): Invoice | null {
+  try {
+    const wh = new Webhook(webhookEndpointSecret);
+    return wh.verify(body, headers) as Invoice;
+  } catch (err) {
+    // Looks like alby might sent multiple webhooks for the same invoice but it should only work once
```
Comment (rolznz) [inline packages/app-store/alby/lib/parseInvoice.ts:17]: This will happen if you disconnect and reconnect the Alby app (the old webhook is not removed and the secret won't match). I will add a TODO for this.

File: packages/app-store/alby/api/add.ts [hunk 11]
```diff
@@ -41,19 +40,3 @@ export default async function handler(req: NextApiRequest, res: NextApiResponse)
 
   return res.status(200).json({ url: "/apps/alby/setup" });
 }
-/*import { createDefaultInstallation } from "@calcom/app-store/_utils/installation";
```
Comment (rolznz) [inline packages/app-store/alby/api/add.ts]: Did we get an answer here? ideally there is no need for a custom handler

File: packages/trpc/server/routers/loggedInViewer/integrations.handler.ts [hunk 12]
```diff
@@ -132,34 +134,48 @@ export const integrationsHandler = async ({ ctx, input }: IntegrationsOptions) =
...
       };
 
+      // We need to know if app is payment type
+      let isSetupAlready = false;
+      if (credential && app.categories.includes("payment")) {
+        const paymentApp = (await appStore[app.dirName as keyof typeof appStore]()) as PaymentApp | null;
+        if (paymentApp && "lib" in paymentApp && paymentApp?.lib && "PaymentService" in paymentApp?.lib) {
+          const PaymentService = paymentApp.lib.PaymentService;
+          const paymentInstance = new PaymentService(credential);
+          isSetupAlready = paymentInstance.isSetupAlready();
```
Comment (rolznz) [inline packages/trpc/server/routers/loggedInViewer/integrations.handler.ts:175]: Cool!

File: packages/trpc/server/routers/loggedInViewer/integrations.handler.ts [hunk 13]
```diff
@@ -170,9 +186,9 @@ export const integrationsHandler = async ({ ctx, input }: IntegrationsOptions) =
         teams,
         isInstalled: !!userCredentialIds.length || !!teams.length || app.isGlobal,
         // FIXME: remove hardcoding and add per-app validation
-        isSetup: !!(credential?.key as { account_id: string })?.account_id || app.slug !== "alby",
+        isSetupAlready,
```
Comment (rolznz) [inline packages/trpc/server/routers/loggedInViewer/integrations.handler.ts:188]: I will remove the FIXME comment above

File: packages/app-store/alby/components/AlbyPaymentComponent.tsx [hunk 14]
```diff
@@ -137,22 +134,17 @@ function PaymentChecker(props: PaymentCheckerProps) {
         if (props.booking.status === "ACCEPTED") {
           return;
         }
-        const bookingsResult = await utils.viewer.bookings.get.fetch({
-          filters: {
-            status: "upcoming",
-            eventTypeIds: [props.eventType.id],
-          },
+        const bookingsResult = await utils.viewer.bookings.find.fetch({
```
Comment (rolznz) [inline packages/app-store/alby/components/AlbyPaymentComponent.tsx:137]: Is it possible to update this to a subscription so the polling can be removed?

File: apps/web/pages/apps/[slug]/setup.tsx [hunk 15]
```diff
@@ -36,11 +36,4 @@ export default function SetupInformation(props: InferGetStaticPropsType<typeof g
...
 
-export const getStaticPaths: GetStaticPaths = async () => {
-  return {
-    paths: [],
-    fallback: "blocking",
-  };
-};
-
-export { getStaticProps };
+export { getServerSideProps };
```
Comment (hariombalhara) [inline apps/web/pages/apps/[slug]/setup.tsx:39]: What's the motivation for this change?

Comment (hariombalhara) [reply apps/web/pages/apps/[slug]/setup.tsx:39]: Got it. Alby needs to do SSR which is a first I guess.

File: packages/features/ee/payments/api/alby-webhook.ts [hunk 16]
```diff
@@ -0,0 +1,125 @@
+import type { NextApiRequest, NextApiResponse } from "next";
```
Comment (hariombalhara) [inline packages/features/ee/payments/api/alby-webhook.ts:1]: nit: I think we should not have app specific files in core. We should plan to move out paypal as well as alby webhook files to their respective apps only

File: packages/app-store/alby/components/AlbyPaymentComponent.tsx [hunk 17]
```diff
@@ -137,22 +134,17 @@ function PaymentChecker(props: PaymentCheckerProps) {
...
+          bookingUid: props.booking.uid,
         });
-        if (
-          bookingsResult.bookings.some(
-            (booking) => booking.id === props.booking.id && booking.status === "ACCEPTED"
-          )
-        ) {
+
+        if (bookingsResult.paid) {
```
Comment (rolznz) [inline packages/app-store/alby/components/AlbyPaymentComponent.tsx]: should be `bookingsResult.booking.paid`

File: apps/web/pages/apps/[slug]/setup.tsx (hunk 15 above)
Comment (rolznz) [reply apps/web/pages/apps/[slug]/setup.tsx:39]: The Alby setup page needs to retrieve app credentials to check if the app has been setup or not.

Possibly it could be done with trpc instead? But, I am wondering do the setup pages make sense to use getStaticProps in the first place if they are for installed apps that require a logged in user?

File: packages/app-store/alby/components/AlbyPaymentComponent.tsx (hunk 14 above)
Comment (alannnc) [reply packages/app-store/alby/components/AlbyPaymentComponent.tsx:138]: We don't have something for subscription, we can use trpc but still revalidates and it will be the same as the polling we are doing right now, probably we can increase the base interval and stop it after 60 seconds?

File: packages/app-store/alby/api/add.ts (hunk 11 above)
Comment (alannnc) [reply packages/app-store/alby/api/add.ts]: My guess here is normal apps doesn't require a handler but since this is a payment and requires setup we should go for the custom one.

File: packages/app-store/alby/lib/parseInvoice.ts (hunk 10 above)
Comment (alannnc) [reply packages/app-store/alby/lib/parseInvoice.ts:17]: Yeah, removing old webhook on disconnect.

File: packages/app-store/alby/components/AlbyPaymentComponent.tsx (hunk 14 above)
Comment (rolznz) [reply packages/app-store/alby/components/AlbyPaymentComponent.tsx:138]: I don't think it would be good to stop polling so quickly. A user might have trouble paying with one wallet and try another for example. But I guess we could slowly ramp up the delay after multiple polls though?

**Comment by PeerRich** [comment]
@rolznz  we're dealing with a few delays here because of the non-alby changes done in this PR.

if you wanna speed it up, we can split this PR into two PRs, one alby PR and one generic app store changes. up to you. we will review this PR regardless right now

File: .env.appStore.example [hunk 18]
```diff
@@ -125,5 +125,11 @@ SALESFORCE_CONSUMER_SECRET=""
 ZOHOCRM_CLIENT_ID=""
 ZOHOCRM_CLIENT_SECRET=""
 
+# ALBY
+# Used for the Alby payment app / receiving Alby payments
+# Get it from: https://getalby.com/developer/oauth_clients
+# Set callbackUrl to YOUR_APP_DOMAIN/api/integrations/alby/alby-webhook
+NEXT_PUBLIC_ALBY_CLIENT_ID=""
+NEXT_PUBLIC_ALBY_CLIENT_SECRET=""
```
Comment (hariombalhara) [inline .env.appStore.example:133]: We don't need these env variables now. See, https://github.com/calcom/cal.com/blob/main/packages/app-store/CONTRIBUTING.md#others

File: apps/web/pages/api/integrations/alby/webhook.ts [hunk 19]
```diff
@@ -0,0 +1 @@
+export { default, config } from "@calcom/features/ee/payments/api/alby-webhook";
```
Comment (hariombalhara) [inline apps/web/pages/api/integrations/alby/webhook.ts:1]: Since it's just an import from there, could we just move the file in here directly. That way, we don't keep alby related code in core.

File: packages/app-store/alby/lib/PaymentService.ts [hunk 20]
```diff
@@ -0,0 +1,133 @@
...
+    bookingId: number,
+    _bookerEmail: string,
+    paymentOption: PaymentOption
+  ): Promise<Payment> {
+    throw new Error("Method not implemented");
+  }
+  chargeCard(
+    payment: Pick<Prisma.PaymentUncheckedCreateInput, "amount" | "currency">,
+    bookingId: number
+  ): Promise<Payment> {
```
Comment (hariombalhara) [inline packages/app-store/alby/lib/PaymentService.ts:105]: We should fix these eslint warnings. I am okay ignoring these warnings directly per line.

Comment (hariombalhara) [reply packages/app-store/alby/lib/PaymentService.ts:105]: Also, I am not sure we really need these definitions in there if they are just noop. Maybe we can remove these methods.

File: yarn.lock
Comment (hariombalhara) [inline yarn.lock:1]: It seems to be have many unrelated changes. Fixing it.

File: packages/app-store/alby/lib/PaymentService.ts (hunk 20 above)
Comment (rolznz) [reply packages/app-store/alby/lib/PaymentService.ts:105]: the method is implemented for stripe, and is part of the interface. The methods could be made optional there, or I can just add ignores on those lines in the Alby implementation

File: packages/app-store/alby/lib/parseInvoice.ts (hunk 10 above)
Comment (rolznz) [reply packages/app-store/alby/lib/parseInvoice.ts:17]: Currently there are some issues on our side here. We can store an OAuth refresh/access token for the app but it'll expire after 30 days and we can only refresh it if there's a booking within that period. If there isn't, we won't be able to call the endpoint to remove the webhook because we won't have a valid token.

For now it seems best to just leave the webhook there and not deal with that additional complexity in the initial version (if the user removes or reinstalls the app, the old webhook calls won't pass the endpoint secret checks and just log an error, but not break anything).

File: .env.appStore.example (hunk 18 above)
Comment (rolznz) [reply .env.appStore.example:133]: I will fix this tomorrow

File: packages/features/ee/payments/api/alby-webhook.ts (hunk 16 above)
Comment (rolznz) [reply packages/features/ee/payments/api/alby-webhook.ts]: looks like this was done for Alby and Paypal, just stripe left to do. I will resolve this for now

File: apps/web/public/static/locales/en/common.json [hunk 21]
```diff
@@ -2059,5 +2059,6 @@
   "edit_users_availability":"Edit user's availability: {{username}}",
   "resend_invitation": "Resend invitation",
   "invitation_resent": "The invitation was resent.",
+  "this_app_is_not_setup_already": "This app has not been setup yet",
```
Comment (rolznz) [inline apps/web/public/static/locales/en/common.json:2062]: nit: should it be `"this_app_is_not_setup_yet"`?

File: .env.appStore.example [hunk 22]
```diff
@@ -125,11 +125,4 @@ SALESFORCE_CONSUMER_SECRET=""
 ZOHOCRM_CLIENT_ID=""
 ZOHOCRM_CLIENT_SECRET=""
 
-# ALBY
-# Used for the Alby payment app / receiving Alby payments
-# Get it from: https://getalby.com/developer/oauth_clients
-# Set callbackUrl to YOUR_APP_DOMAIN/api/integrations/alby/alby-webhook
-NEXT_PUBLIC_ALBY_CLIENT_ID=""
```
Comment (rolznz) [inline .env.appStore.example]: These were removed but not from turbo.json and `packages/app-store/alby/pages/setup/index.tsx` needs to be updated to use the client_id and client_secret properties

File: packages/trpc/server/routers/viewer/eventTypes/update.handler.ts [hunk 23]
```diff
@@ -240,49 +240,31 @@ export const updateHandler = async ({ ctx, input }: UpdateOptions) => {
...
-    if (paymentCredential?.appId === "stripe") {
-      const { default_currency } = stripeDataSchema.parse(paymentCredential.key);
-      data.currency = default_currency;
-    }
+  /**
+   * Since you can have multiple payment apps we will honor the first one to save in eventType
+   * but the real detail will be inside app metadata, so with this you can have different prices in different apps
+   * So the price and currency inside eventType will be deprecated soon or just keep as reference.
+   */
+  if (
```
Comment (rolznz) [inline packages/trpc/server/routers/viewer/eventTypes/update.handler.ts:248]: This has exposed a bug. I only have Alby installed (stripe is not installed at https://endless-close-ocelot.ngrok-free.app/apps/stripe and is not listed in my apps on the event type), but `input.metadata.apps` also contains stripe and therefore the stripe price and currency are picked since they are checked first. @alannnc can you help here?

I think how this works is a bit broken - either it should be enforced that there can only be one payment app per event type (probably best for now?) or add proper support for accepting multiple payment methods and do not store a `price` field on the event type.

```
"metadata": {
    "apps": {
      "alby": {
        "enabled": true,
        "price": 1,
        "currency": "BTC"
      },
      "giphy": {
        "enabled": false,
        "thankYouPage": ""
      },
      "stripe": {
        "enabled": true,
        "price": 1,
        "currency": "usd"
      }
    }
  },
```

image

File: apps/web/pages/api/integrations/alby/webhook.ts (hunk 19 above)
Comment (rolznz) [reply apps/web/pages/api/integrations/alby/webhook.ts]: this was fixed by Alan

File: packages/app-store/alby/lib/PaymentService.ts (hunk 20 above)
Comment (rolznz) [reply packages/app-store/alby/lib/PaymentService.ts:105]: I fixed the warnings for now by adding `_` prefix on the arguments

File: .env.appStore.example (hunk 22 above)
Comment (rolznz) [reply .env.appStore.example]: fixed

File: packages/trpc/server/routers/viewer/eventTypes/update.handler.ts (hunk 23 above)
Comment (alannnc) [reply packages/trpc/server/routers/viewer/eventTypes/update.handler.ts:248]: Yeah I ended in the same path, looks like legacy bug since it also saves giphy which shouldn't be the case. My quick fix was changing order of paymet apps so it would take alby | paypal first before stripe.

Comment (rolznz) [reply packages/trpc/server/routers/viewer/eventTypes/update.handler.ts:248]: @alannnc so you already fixed it locally? can you push the fix?

Comment (alannnc) [reply packages/trpc/server/routers/viewer/eventTypes/update.handler.ts:248]: Should be fixed now.

**Review by alannnc — DISMISSED** [review]
Looks like should be ready now. :)

**Review by PeerRich — APPROVED** [review]
reapproved on behalf of @alannnc
